george69/Interpreter.py

```diff
@@ -0,0 +1,389 @@
+            self.mem = bytearray(fp.read())
+    def executeProgram(self):
+        try:
+            while True:
+                instruction = Instruction(self.mem[self.pc])
+                ##print(instruction.numericalInstruction)
+                #print(instruction.instruction)
+                #print(self.pc)
+                fun = self.availableInstructions.get(instruction.instruction)
+                fun()
+                #print([str(d) for d in self.rstack])
+                if instruction.instruction == "halt":
+                    break
+        except Exception as e:
+            print("ERROR OCCURED, DUMPING")
+            print(e)
+            self.halt()
+
+    def halt(self):
+        print()
+        print("Compile values:")
+        print(str(self), end="")
+
+    def call(self):
+        #self.halt()
+        self.fpsp += 1
+        self.fpstack.insert(self.fpsp, self.sp - self.rstack[self.sp].value - 1)
+        del self.rstack[self.sp]
+        self.sp -= 1
+        self.pc = self.rstack[self.sp].value
+        #print("GOTO " + str(self.rstack[self.sp].value))
+        del self.rstack[self.sp]
+        #print(str([str(d.value) for d in self.rstack]))
+        #print("PC: " + str(self.pc))
+        #del self.rstack[self.sp]
+        self.sp -= 1
+        #print(self.fpstack)
+
+    def ret(self):
+        #print([str(d) for d in self.rstack])
+        self.sp = self.fpstack[self.fpsp]
+        #print(self.sp)
+        del self.fpstack[self.fpsp]
+        self.fpsp -= 1
+        self.pc = self.rstack[self.sp].value
+        del self.rstack[self.sp]
+        self.sp -= 1
+
+    def jmp(self):
+        self.pc = self.rstack[self.sp].value
+        del self.rstack[self.sp]
+        self.sp -= 1
+
+    def jmpc(self):
+        if self.rstack[self.sp - 1].value == 1:
+            self.pc = self.rstack[self.sp].value
+        else:
+            self.pc += 1
+        del self.rstack[self.sp]
+        del self.rstack[self.sp - 1]
+        self.sp -= 2
+
+    def pushc(self):
+        data = Datum(self.mem[self.pc+1], "char")
+        self.sp += 1
+        self.rstack.insert(self.sp, data)
+        self.pc += data.size
+        self.pc += 1
+
+    def pushs(self):
+        data = Datum(int.from_bytes(bytearray(self.mem[self.pc + 1:self.pc + 3]), byteorder="little"), "short")
+        self.sp += 1
+        self.rstack.insert(self.sp, data)
+        self.pc += data.size
+        self.pc += 1
+
+    def pushi(self):
+        data = Datum(int.from_bytes(bytearray(self.mem[self.pc + 1:self.pc + 5]), byteorder="little"), "int")
+        #print(data.value)
+        self.sp += 1
+        self.rstack.insert(self.sp, data)
+        self.pc += data.size
+        self.pc += 1
+
+    def pushf(self):
+        data = Datum(struct.unpack("<f", bytearray(self.mem[self.pc+1:self.pc + 5]))[0], "float")
+        self.sp += 1
+        self.rstack.insert(self.sp, data)
+        self.pc += data.size
+        self.pc += 1
+
+    def pushvi(self):
```

Comment:
If code is the same condense these into a singular function and have pushvi, pushvc, pushvs, and pushfv all call that one function.